### PR TITLE
Feature/event listener

### PR DESCRIPTION
**Pull request by thomaszurkan-optimizely**

### CODE REVIEW

File: optimizely/notification_center.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+class NotificationCenter(object):
+  """ Class encapsulating Broadcast Notifications. The enums.NotifcationTypes includes predefined notifications."""
+  def __init__(self, logger):
+    self.notification_id = 1
+    self.notifications = {}
+    for attr, value in enums.NotificationTypes.__dict__.items():
+      self.notifications[value] = []
+    self.logger = logger
+
+  def add_notification(self, notification_type, notification_callback):
```
Comment (mikeng13) [inline optimizely/notification_center.py]: Can we call this `add_notification_listener` instead? It more accurately represents what the API does.

File: optimizely/notification_center.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+          return -1
+      self.notifications[notification_type].append((self.notification_id, notification_callback))
+
+    ret_val = self.notification_id
+
+    self.notification_id += 1
+
+    return ret_val
+
+  def remove_notification(self, notification_id):
```
Comment (mikeng13) [inline optimizely/notification_center.py]: Same here: `remove_notification_listener`

File: optimizely/notification_center.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      notification_type: .helpers.enums.NotificationTypes
+      notification_callback: closure of function to call when event is triggered.
+    Returns:
+      notification id used to remove the notification
+    """
+    if notification_type not in self.notifications:
+      self.notifications[notification_type] = [(self.notification_id, notification_callback)]
+    else:
+      for callback in self.notifications[notification_type]:
+        if callback[1] == notification_callback:
```
Comment (mikeng13) [inline optimizely/notification_center.py]: A cleaner way of accessing the callback would be to unpack the tuple in the `for` declaration like this:

`for notification_id, callback in self.notifications[notification_type]:`

File: optimizely/notification_center.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  def fire_notifications(self, notification_type, *args):
+    """ Fires off the notification for the specific event.  Uses var args to pass in a
+        arbitrary list of parameter according to which notification type was fired.
+
+    Args:
+      notification_type: Type of notification to fire.
+      args: list of arguments to the callback.
+    """
+    if notification_type in self.notifications:
+      for callback in self.notifications[notification_type]:
```
Comment (mikeng13) [inline optimizely/notification_center.py]: Same as above, unpack the tuple here so you can just call `callback(*args)`

File: optimizely/optimizely.py [hunk 5]
```diff
@@ -316,7 +320,8 @@ def track(self, event_key, user_id, attributes=None, event_tags=None):
       except:
         error = sys.exc_info()[1]
         self.logger.log(enums.LogLevels.ERROR, 'Unable to dispatch conversion event. Error: %s' % str(error))
-
+      self.notification_center.fire_notifications(enums.NotificationTypes.TRACK, event_key, user_id,
```
Comment (mikeng13) [inline optimizely/optimizely.py]: This actually is interesting. Do we fire the notification as soon as we dispatch the event, or we wait for the logendpoint to return to us before we tell the listener? I could see us running into counting issues (mismatches) between us and other analytics providers if we pass the event without first guaranteeing that we received it on our end. Thoughts @aliabbasrizvi @wangjoshuah @ceimaj ?

Comment (thomaszurkan-optimizely) [reply optimizely/optimizely.py]: IMHO, I think if there is a mismatch then there would be a problem on our end and it would probably bubble up faster.

Comment (mikeng13) [reply optimizely/optimizely.py]: Yeah I don't feel too strongly about it. The key here is to very clearly document in our dev docs, WHEN exactly we fire off the notification.

Comment (thomaszurkan-optimizely) [reply optimizely/optimizely.py]: well, that actually brings up a good point, do we want to have pre and post calls?

Comment (mikeng13) [reply optimizely/optimizely.py]: nah, I think that gets too complicated. The current approach you have there should be good.

File: optimizely/helpers/enums.py [hunk 6]
```diff
@@ -41,3 +41,13 @@ class Errors(object):
   INVALID_VARIATION_ERROR = 'Provided variation is not in datafile.'
   UNSUPPORTED_DATAFILE_VERSION = 'Provided datafile has unsupported version. ' \
                                  'Please use SDK version 1.1.0 or earlier for datafile version 1.'
+
+
+class NotificationTypes(object):
+  """ NotificationTypes for the notification_center.NotificationCenter
+      format is NOTIFICATION TYPE: list of parameters to callback.
+  """
+  ACTIVATE = "ACTIVATE:experiment, user_id,attributes, variation, event"
```
Comment (mikeng13) [inline optimizely/helpers/enums.py]: What was the reason for changing this back to `ACTIVATE` from `DECISION`? `ACTIVATE` is more of an Optimizely term and might not make sense to developers of other anaytics products. Instead `DECISION` is a more industry-standard term.

Comment (mikeng13) [reply optimizely/helpers/enums.py]: @thomaszurkan-optimizely I left a comment here in case you miss it

File: optimizely/optimizely.py [hunk 7]
```diff
@@ -383,6 +388,19 @@ def is_feature_enabled(self, feature_key, user_id, attributes=None):
...
                                     attributes)
+        self.notification_center.send_notifications(enums.NotificationTypes.FEATURE_EXPERIMENT,
+                                                  feature_key, user_id, attributes,
+                                                    decision.experiment, decision.variation)
+      else:
+        audience = None
+        if decision.experiment is not None and \
+                decision.experiment.audienceIds is not None and \
+                len(decision.experiment.audienceIds) > 0:
+          audience_id = decision.experiment.audienceIds[0]
```
Comment (oakbani) [inline optimizely/optimizely.py]: @thomaszurkan-optimizely Hi, I have a question here. In the case where the rollout rule/experiment has multiple audience IDs, and the user didn't qualify for the first audience, but for some other audience id in the list. Do we still send the audience at index 0 in send_notifications method?

**Pull request merged**